src/Internal/Service/DuplicatePaymentPreventionService.php

```diff
@@ -0,0 +1,260 @@
+	private $order_service;
+
+	/**
+	 * HooksProxy instance.
+	 *
+	 * @var HooksProxy
+	 */
+	private $hooks_proxy;
+
+	/**
+	 * LegacyProxy instance.
+	 *
+	 * @var LegacyProxy
+	 */
+	private $legacy_proxy;
+
+	/**
+	 * Session service instance.
+	 *
+	 * @var SessionService
+	 */
+	private $session_service;
+
+	/**
+	 * Logger instance.
+	 *
+	 * @var Logger
+	 */
+	private $logger;
+
+	/**
+	 * Initializes all dependencies.
+	 *
+	 * @param  OrderService   $order_service   Order service instance.
+	 * @param  SessionService $session_service Session service instance.
+	 * @param  Logger         $logger          Logger instance.
+	 * @param  HooksProxy     $hooks_proxy     Hooks proxy instance.
+	 * @param  LegacyProxy    $legacy_proxy    Legacy proxy instance.
+	 */
+	public function __construct( OrderService $order_service, SessionService $session_service, Logger $logger, HooksProxy $hooks_proxy, LegacyProxy $legacy_proxy ) {
+		$this->order_service   = $order_service;
+		$this->session_service = $session_service;
+		$this->logger          = $logger;
+		$this->hooks_proxy     = $hooks_proxy;
+		$this->legacy_proxy    = $legacy_proxy;
+	}
+
+	/**
+	 * Initializes this class's hooks.
+	 *
+	 * @return void
+	 */
+	public function init_hooks(): void {
+		// Priority 21 to run right after wc_clear_cart_after_payment.
+		$this->hooks_proxy->add_action( 'template_redirect', [ $this, 'clear_session_processing_order_after_landing_order_received_page' ], 21 );
+	}
+
+	/**
+	 * Checks if the currently attached payment intent was authorized for the current processing order.
+	 *
+	 * @param  int $order_id ID of the current processing order.
+	 *
+	 * @return WC_Payments_API_Payment_Intention|null The authorized attached payment intent, null otherwise.
+	 * @throws Order_Not_Found_Exception
+	 */
+	public function get_authorized_payment_intent_attached_to_order( int $order_id ): ?WC_Payments_API_Payment_Intention {
+		$intent_id = $this->order_service->get_intent_id( $order_id );
+		if ( is_null( $intent_id ) ) {
+			return null;
+		}
+
+		// We only care about payment payment_intent.
+		$is_payment_intent = 'pi_' === substr( $intent_id, 0, 3 );
+		if ( ! $is_payment_intent ) {
+			return null;
+		}
+
+		try {
+			$request = Get_Intention::create( $intent_id );
+			$request->set_hook_args( $this->order_service->_deprecated_get_order( $order_id ) );
+			/** @var WC_Payments_API_Payment_Intention $payment_intent */ // phpcs:ignore Generic.Commenting.DocComment.MissingShort
+			$payment_intent = $request->send();
+		} catch ( \Exception $e ) {
+			$this->logger->error( 'Failed to fetch attached payment intent: ' . $e );
+			return null;
+		};
+
+		if ( ! $payment_intent->is_authorized() ) {
+			return null;
+		}
+
+		$intent_meta_order_id_raw = $payment_intent->get_metadata()['order_id'] ?? '';
+		$intent_meta_order_id     = is_numeric( $intent_meta_order_id_raw ) ? intval( $intent_meta_order_id_raw ) : 0;
+		if ( $intent_meta_order_id !== $order_id ) {
```

Comment:
Question: When would this ever happen? I know it was there before, but I don't remember.

I'm asking, because maybe we should kill the payment process, rather than allowing the code to proceed here?

Reply:
When I wrote the comment, I was considering clearing the cart, and forcing the customer into a new order. However, if we do that, it will probably make more sense as a general part of the payment process, not something deep within the DPPS. Marking this as resolved.